p2p/resolve_ref.go

```diff
@@ -2,88 +2,186 @@ package p2p
+	"time"
 
-	"github.com/qri-io/qri/repo"
-	reporef "github.com/qri-io/qri/repo/ref"
+	"github.com/libp2p/go-libp2p-core/helpers"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	protocol "github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/qri-io/qri/dsref"
 )
 
-// MtResolveDatasetRef resolves a dataset reference
-const MtResolveDatasetRef = MsgType("resolve_dataset_ref")
+const (
+	// p2pRefResolverTimeout is the length of time we will wait for a
+	// RefResolverRequest response before cancelling the context
+	// this can potentially be a config option in the future
+	p2pRefResolverTimeout = time.Second * 60
+	// ResolveRefProtocolID is the protocol on which qri nodes communicate to
+	// resolve references
+	ResolveRefProtocolID = protocol.ID("/qri/ref/0.1.0")
+)
 
-// ResolveDatasetRef completes a dataset reference
-func (n *QriNode) ResolveDatasetRef(ctx context.Context, ref *reporef.DatasetRef) (err error) {
-	log.Debugf("%s ResolveDatasetRef %s", n.ID, ref)
+type p2pRefResolver struct {
+	node *QriNode
+}
 
-	if !n.Online {
-		return ErrNotConnected
-	}
+type resolveRefRes struct {
+	ref    *dsref.Ref
+	source string
+}
 
-	pids := n.ClosestConnectedQriPeers(ref.ProfileID, 15)
-	if len(pids) == 0 {
-		return fmt.Errorf("no connected peers")
+func (rr *p2pRefResolver) ResolveRef(ctx context.Context, ref *dsref.Ref) (string, error) {
+	log.Debugf("p2p.ResolveRef ref=%q", ref)
+	if rr == nil || rr.node == nil {
+		return "", dsref.ErrRefNotFound
 	}
+	refCp := ref.Copy()
+	streamCtx, cancel := context.WithTimeout(ctx, p2pRefResolverTimeout)
+	defer cancel()
 
-	replies := make(chan Message)
-	req, err := NewJSONBodyMessage(n.ID, MtResolveDatasetRef, ref)
-	req = req.WithHeaders("phase", "request")
-	if err != nil {
-		log.Debug(err.Error())
-		return err
+	connectedPids := rr.node.ConnectedQriPeerIDs()
+	numReqs := len(connectedPids)
+	if numReqs == 0 {
+		return "", dsref.ErrRefNotFound
 	}
 
-	for _, pid := range pids {
-		if err := n.SendMessage(ctx, req, replies, pid); err != nil {
-			log.Debugf("%s err: %s", pid, err.Error())
-			continue
-		}
+	refMu := sync.Mutex{}
+	resCh := make(chan resolveRefRes, numReqs)
+	for _, pid := range connectedPids {
+		refMu.Lock()
+		reqRef := refCp.Copy()
+		refMu.Unlock()
+		go func(pid peer.ID, reqRef *dsref.Ref) {
+			source := rr.resolveRefRequest(streamCtx, pid, reqRef)
+			resCh <- resolveRefRes{
+				ref:    reqRef,
+				source: source,
+			}
+		}(pid, &reqRef)
+	}
 
-		res := <-replies
-		dsr := reporef.DatasetRef{}
-		if err := json.Unmarshal(res.Body, &dsr); err == nil {
-			if dsr.Path != "" {
-				*ref = dsr
-				break
+	for {
+		select {
+		case res := <-resCh:
+			numReqs--
+			if !res.ref.Complete() && numReqs == 0 {
+				return "", dsref.ErrRefNotFound
+			}
+			if res.ref.Complete() {
+				*ref = *res.ref
+				return res.source, nil
 			}
+		case <-streamCtx.Done():
+			log.Debug("p2p.ResolveRef request timed out before resolving ref")
```

Comment:
this might not be a timeout. the parent context might be cancelled. I say this because that's really what we should be doing in the long run: creating a sub context that's passed to all parallel resolvers and then cancelling when any of them returns with a resolved reference.

This way, if the registry is in a parallel resolver with the p2p resolver, and it comes back first, all this p2p work will be cleaned up faster